pkg/mux/mux.go

```diff
@@ -0,0 +1,272 @@
+package mux
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/bitfinexcom/bitfinex-api-go/pkg/models/event"
+	"github.com/bitfinexcom/bitfinex-api-go/pkg/mux/client"
+	"github.com/bitfinexcom/bitfinex-api-go/pkg/mux/msg"
+)
+
+// Mux will manage all connections and subscriptions. Will check if subscriptions
+// limit is reached and spawn new connection when that happens. It will also listen
+// to all incomming client messages and reconnect client with all its subscriptions
+// in case of a failure
+type Mux struct {
+	cid           int
+	dms           int
+	publicChan    chan msg.Msg
+	publicClients map[int]*client.Client
+	privateChan   chan msg.Msg
+	privateClient *client.Client
+	mtx           *sync.RWMutex
+	Err           error
+	transform     bool
+	apikey        string
+	apisec        string
+	subInfo       map[int64]event.Info
+	authenticated bool
+	publicURL     string
+	authURL       string
+}
+
+// New returns pointer to instance of mux
+func New() *Mux {
+	return &Mux{
+		publicChan:    make(chan msg.Msg),
+		privateChan:   make(chan msg.Msg),
+		publicClients: make(map[int]*client.Client),
+		mtx:           &sync.RWMutex{},
+		subInfo:       map[int64]event.Info{},
+		publicURL:     "wss://api-pub.bitfinex.com/ws/2",
+		authURL:       "wss://api.bitfinex.com/ws/2",
+	}
+}
+
+// TransformRaw enables data transformation and mapping to appropriate
+// models before sending it to consumer
+func (m *Mux) TransformRaw() *Mux {
+	m.transform = true
+	return m
+}
+
+// WithAPIKEY accepts and persists api key
+func (m *Mux) WithAPIKEY(key string) *Mux {
+	m.apikey = key
+	return m
+}
+
+// WithDeadManSwitch - when socket is closed, cancel all account orders
+func (m *Mux) WithDeadManSwitch() *Mux {
+	m.dms = 4
+	return m
+}
+
+// WithAPISEC accepts and persists api sec
+func (m *Mux) WithAPISEC(sec string) *Mux {
+	m.apisec = sec
+	return m
+}
+
+// WithAuthURL accepts and persists auth url
+func (m *Mux) WithAuthURL(url string) *Mux {
+	m.authURL = url
+	return m
+}
+
+// Subscribe - given the details in form of event.Subscribe,
+// subscribes client to public channels
+func (m *Mux) Subscribe(sub event.Subscribe) *Mux {
+	if m.Err != nil {
+		return m
+	}
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	if subscribed := m.publicClients[m.cid].SubAdded(sub); subscribed {
+		return m
+	}
+
+	if m.Err = m.publicClients[m.cid].Subscribe(sub); m.Err != nil {
+		return m
+	}
+
+	if limitReached := m.publicClients[m.cid].SubsLimitReached(); limitReached {
```

Comment:
unlikely in most prod scenarios, but i think this can happen, imagine a subscription limit of 2:

1. subscribe to btcusd, btceth
2. subscribe to ethusd, ethusdt
3. subscribe to eosusd (we have 3 clients running, as subscription limit is 1)
4. unsubscribe btcusd
5. subscribe foousd
6. open clients == 4, one with just one subscription?

Reply:
not a blocker, but maybe an "advanced feature" we can add later if my observation was right